crates/async-compression/src/generic/bufread/decoder.rs

```diff
@@ -0,0 +1,126 @@
+use crate::codecs::Decode;
+use crate::core::util::PartialBuffer;
+
+use core::task::{Context, Poll};
+use std::io::Result;
+
+use futures_core::ready;
+
+#[derive(Debug)]
+enum State {
+    Decoding,
+    Flushing,
+    Done,
+    Next,
+}
+
+pub(crate) trait AsyncBufRead {
+    fn poll_fill_buf(&mut self, cx: &mut Context<'_>) -> Poll<Result<&[u8]>>;
+    fn consume(&mut self, bytes: usize);
```

Comment:
Given that the reader is already pinned, using `&mut` is much simpler here for this wrapper method